Unify metrics

Andrewwango opened the pull request:
Closes #299. Unify metrics in one submodule. Note there are a lot of changed files from metrics being moved from `utils` into `loss.metric`.

- Add `Metric` baseclass that all metrics should inherit from
- All metrics can perform preprocessing (complex abs, norm) and postprocessing (training, reduce)
- Metrics can be used either as metrics, or as losses (e.g. replacing `torch.nn.MSELoss()`)
- Unit tests for metrics
- Add L1L2 metric


### Checks to be done before submitting your PR
- [x] `python3 -m pytest tests/` runs successfully.
- [x] `black .` runs successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

tachella commented on deepinv/loss/metric/metric.py:
```diff
@@ -0,0 +1,158 @@
+        )
+
+
+class Metric(Loss):
```
Do we want to inherit from Loss?

I am afraid that we end up with some circular patterns, eg EILoss using a metric

Andrewwango replied:
Metric now no longer inherits from loss

tachella commented on deepinv/loss/metric/metric.py:
```diff
@@ -0,0 +1,158 @@
+
+    :param bool complex_abs: perform complex magnitude before passing data to metric function. If ``True``,
+        the data must either be of complex dtype or have size 2 in the channel dimension (usually the second dimension after batch).
+    :param bool train_loss: use metric as a training loss, by returning one minus the metric. If lower is better, does nothing.
```
I'm not sure if we should assume that metrics are always between 0 and 1, with 1 being bad.
I would remove this option.

Andrewwango replied:
I've now changed it such that the train_loss option is only visible in the docs for SSIM, however I have kept it as an option for all metrics, so that it's very easy to implement for future metrics which would also require it. I have generalised it by adding a invert_metric method which can be overridden (e.g. for SSIM, it should be 1-m). I've made this clear in the docstrings.

tachella commented on deepinv/loss/metric/perceptual.py:
```diff
@@ -0,0 +1,77 @@
+
+    >>> from deepinv.utils.demo import get_image_url, load_url_image
+    >>> from deepinv.loss.metric import LPIPS
+    >>> ();m = LPIPS();() # doctest: +ELLIPSIS
```
what is this line doing?

Andrewwango replied:
It's because pyiqa is annoyingly written (they use conditional print statements in their code instead of warnings) + doctest is a bit fiddly so this is the only hacky way to capture the print output

tachella commented on deepinv/loss/scheduler.py:
```diff
@@ -89,8 +89,9 @@ class RandomLossScheduler(BaseLossScheduler):
-    >>> l = RandomLossScheduler(SupLoss(), SSIM()) # Choose randomly between Sup and SSIM
+    >>> from deepinv.loss import RandomLossScheduler, SupLoss
+    >>> from deepinv.loss.metric import SSIM
+    >>> l = RandomLossScheduler(SupLoss(), SSIM(train_loss=True)) # Choose randomly between Sup and SSIM
```
see issue with train_loss option above - I'm okay with having a special kwarg for SSIM

tachella commented on deepinv/tests/test_metric.py:
```diff
@@ -0,0 +1,152 @@
+    x_hat = GaussianNoise(sigma=0.1, rng=rng)(x)
+
+    # Test metric worse when image worse
+    if train_loss:  # All metrics lower = better
```
some metrics can have higher=better, right?

Andrewwango replied:
i've added a comment to explain this in the test

tachella commented on deepinv/training/trainer.py:
```diff
@@ -137,7 +138,7 @@ def get_samples_offline(self, iterators, g):
         the physics operators. If not None, the physics operators are randomly sampled at each iteration using the generator.
         Should be used in conjunction with ``online_measurements=True``. Also see ``loop_physics_generator``.
-    :param deepinv.loss.Loss, list[deepinv.loss.Loss] metrics: Metric or list of metrics used for evaluating the model.
+    :param deepinv.loss.metric.Metric, list[deepinv.loss.metric.Metric] metrics: Metric or list of metrics used for evaluating the model.
```
can we have deepinv.loss.Metric? otherwise I feel is a bit verbose

Andrewwango replied:
You can either import from `deepinv.loss.Metric`, `deepinv.metric.Metric`, or `deepinv.loss.metric.Metric`. These are all tested in the `test_metrics` tests. This means in most cases it remains backwards compatible.

tachella commented on docs/source/deepinv.loss.rst:
```diff
@@ -93,19 +93,58 @@ Training is implemented using :class:`deepinv.training.AdversarialTrainer` which
-Metrics are generally used to evaluate the performance of a model. Some of them can be used as training losses as well.
+Metrics are generally used to evaluate the performance of a model.
+
+Metrics inherit from the :class:`deepinv.loss.metric.Metric` baseclass, and take either ``x_net, x``
```
deepinv.loss.Metric

Andrewwango replied:
See above that you can import either `deepinv.metric.Metric`, `deepinv.loss.Metric`or `deepinv.loss.metric.Metric`. I've added a line explaining this. However annoyingly it doesn't create the link if I shorten it, so it will have to be verbose

Andrewwango replied:
renaming them also emphasises that these are metrics, not normal losses, i.e. you can't use physics and model with them

tachella commented on docs/source/deepinv.loss.rst:
```diff
@@ -93,19 +93,58 @@ Training is implemented using :class:`deepinv.training.AdversarialTrainer` which
-        deepinv.loss.SSIM
-        deepinv.loss.LPIPS
-        deepinv.loss.NIQE
+        deepinv.loss.metric.Metric
```
not sure this renaming is that useful - what is the idea behind?

tachella replied:
I worry that it is not backwards compatible

Andrewwango replied:
see above

tachella commented on examples/plug-and-play/demo_vanilla_PnP.py:
```diff
@@ -134,8 +134,8 @@
 # compute PSNR
-print(f"Linear reconstruction PSNR: {dinv.utils.metric.cal_psnr(x, x_lin):.2f} dB")
-print(f"PnP reconstruction PSNR: {dinv.utils.metric.cal_psnr(x, x_model):.2f} dB")
+print(f"Linear reconstruction PSNR: {dinv.metric.PSNR()(x, x_lin).item():.2f} dB")
```
shouldn't be dinv.loss.metric.PSNR?

Andrewwango replied:
you can import directly with dinv.metric.PSNR, see above

tachella reviewed (commented):
I have some questions regarding the refactoring:

1. Do we want `Metric` to inherit from `Loss`? I'm worried there is some circularity here.
2. I'm not sure it is a good idea to refactor all metrics to `deepinv.loss.metric`. This is not backwards compatible and I'm not sure if it clarifies things much. If go down the non-backwards compatible path, it could directly be `deepinv.metric`.
3. I'm not sure regarding whether all metrics should verify lower = better, eg SSIM and MSE are opposite.

Andrewwango commented:
Thanks for your review @tachella . Re. your comments:
1. I have removed Metric inheriting from Loss, and now Metric expects `(x_net, x=None, y=None)`
2. All metrics can be imported from `deepinv.loss.metric`, `deepinv.metric`, or `deepinv.loss`, so we will have backwards compatibility. For the old functions in `deepinv.utils.metric`, I have kept them and added deprecation errors.
3. `lower_better` is an attribute that should be defined in all implementations of metrics, this is what they do in `pyiqa` too. When `lower_better=True` (most of the metrics), it doesn't affect anything. However when `lower_better=False`, as is the case in SSIM, the inverse is taken when `train_loss=True` is passed. I have made this clearer in the docs + docstrings.

tachella reviewed (changes requested):
Many thanks for the refactoring. 

-  now that metric is a separate class from loss, I would move the metric code to a new folder, outside loss.

- some docstrings refer to the forward method but there is no specific forward method for those classes. I would incorporate a short description of the metric instead.

Andrewwango commented on deepinv/models/dip.py:
```diff
@@ -1,7 +1,7 @@
 import torch.nn as nn
 import numpy as np
-from deepinv.loss import MCLoss
+from deepinv.loss.mc import MCLoss
```
fixed

Andrewwango commented on deepinv/loss/metric/distortion.py:
```diff
@@ -0,0 +1,401 @@
+
+    Note we don't use the torchmetrics implementation.
+
+    See docs for ``forward()`` below for more details.
```
done

Andrewwango commented on deepinv/loss/metric/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from .metric import Metric
```
done

Andrewwango commented on deepinv/loss/metric/distortion.py:
```diff
@@ -0,0 +1,401 @@
+    r"""
+    Mean Squared Error metric.
+
+    See docs for ``forward()`` below for more details.
```
done

Andrewwango commented on deepinv/loss/metric/distortion.py:
```diff
@@ -0,0 +1,401 @@
+
+    Normalises MSE by the L2 norm of the ground truth ``x``.
+
+    See docs for ``forward()`` below for more details.
```
done

Andrewwango replied:
done

Andrewwango commented on deepinv/utils/plotting.py:
```diff
@@ -211,7 +212,8 @@ def plot(
     :param list[torch.Tensor], torch.Tensor img_list: list of images to plot or single image.
     :param list[str] titles: list of titles for each image, has to be same length as img_list.
-    :param None, str, Path save_dir: path to save the plot.
+    :param None, str, Path save_fn: path to save the plot as a single image.
```
I think it's useful, I've added an explanation in the docs
